code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/Unifier.kt

```diff
@@ -0,0 +1,91 @@
+package org.jetbrains.research.code.submissions.clustering.load
+
+import com.intellij.ide.impl.ProjectUtil
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiManager
+import com.jetbrains.python.PythonFileType
+import org.jetbrains.research.code.submissions.clustering.Submission
+import org.jetbrains.research.code.submissions.clustering.util.addPyFileToProject
+import org.jetbrains.research.code.submissions.clustering.util.getTmpProjectDir
+import org.jetbrains.research.ml.ast.transformations.Transformation
+import org.jetbrains.research.ml.ast.transformations.anonymization.AnonymizationTransformation
+import org.jetbrains.research.ml.ast.transformations.augmentedAssignment.AugmentedAssignmentTransformation
+import org.jetbrains.research.ml.ast.transformations.commentsRemoval.CommentsRemovalTransformation
+import org.jetbrains.research.ml.ast.transformations.comparisonUnification.ComparisonUnificationTransformation
+import org.jetbrains.research.ml.ast.transformations.constantfolding.ConstantFoldingTransformation
+import org.jetbrains.research.ml.ast.transformations.deadcode.DeadCodeRemovalTransformation
+import org.jetbrains.research.ml.ast.transformations.expressionUnification.ExpressionUnificationTransformation
+import org.jetbrains.research.ml.ast.transformations.ifRedundantLinesRemoval.IfRedundantLinesRemovalTransformation
+import org.jetbrains.research.ml.ast.transformations.multipleOperatorComparison.MultipleOperatorComparisonTransformation
+import org.jetbrains.research.ml.ast.transformations.multipleTargetAssignment.MultipleTargetAssignmentTransformation
+import org.jetbrains.research.ml.ast.transformations.outerNotElimination.OuterNotEliminationTransformation
+import org.jetbrains.research.ml.ast.util.sdk.setSdkToProject
+
+private const val MAX_ITERATIONS: Int = 5
+
+/**
+ * Class for unifying code solutions using ast transformations.
+ */
+class Unifier {
```

Comment:
Let's make it more complex to add the possibility to work with other languages. Create an interface or an abstract class, where you can store psiManager, uninitialized transformationsList, extension or a language, the implementation for applyTransformations function and so on. If you want we can schedule a call to discuss it in more detail. But the main goal to make it is possible to work with different languages in the future.

Reply:
Let's show a possible refactoring:
```
enum class Language(val extension: String) {
    PYTHON(PythonFileType.INSTANCE.defaultExtension)
}

abstract class AbstractUnifier {
    companion object {
        const val MAX_ITERATIONS: Int = 5
    }

    protected val project: Project = ProjectUtil.openOrImport(getTmpProjectDir(), null, true)
        ?: error("Internal error: the temp project was not created")
    protected val psiManager: PsiManager = PsiManager.getInstance(project)

    abstract val language: Language
    abstract val transformations: List<Transformation>

    private fun String.createPsiFile(id: Int): PsiFile = ApplicationManager.getApplication().runWriteAction<PsiFile> {
        val basePath = getTmpProjectDir(toCreateFolder = false)
        val fileName = "dummy$id.${language.extension}"
        val file = addPyFileToProject(basePath, fileName, fileContext = this)
        val virtualFile = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(file)
        psiManager.findFile(virtualFile!!)
    }

    private fun PsiFile.applyTransformations(transformations: List<Transformation>) {
        ApplicationManager.getApplication().invokeAndWait {
            ApplicationManager.getApplication().runWriteAction {
                transformations.forEach {
                    it.forwardApply(this)
                }
            }
        }
    }

    private fun PsiFile.hasChangedAfterTransformations(transformations: List<Transformation>): Boolean {
        val prevPsi = this.copy()
        this.applyTransformations(transformations)
        return prevPsi.text != this.text
    }

    fun Submission.unify(): Submission {
        val psi = this.code.createPsiFile(this.id)
        var wasChanged: Boolean
        var iterationsCounter = 0
        do {
            wasChanged = psi.hasChangedAfterTransformations(transformations)
            iterationsCounter += 1
        } while (wasChanged || iterationsCounter <= MAX_ITERATIONS)

        return this.copy(code = psi.text)
    }
}

class PyUnifier: AbstractUnifier() {
    override val language = Language.PYTHON
    override val transformations = listOf(
        AnonymizationTransformation,
        AugmentedAssignmentTransformation,
        CommentsRemovalTransformation,
        ComparisonUnificationTransformation,
        ConstantFoldingTransformation,
        DeadCodeRemovalTransformation,
        ExpressionUnificationTransformation,
        IfRedundantLinesRemovalTransformation,
        MultipleOperatorComparisonTransformation,
        MultipleTargetAssignmentTransformation,
        OuterNotEliminationTransformation,
    )

    init {
        setSdkToProject(project, getTmpProjectDir())
    }
}
```

Reply:
Also, please add tests for python unifier